**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: David G. Pérez Sifuentes
*   MATRÍCULA: A01541553


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [20]:
#python -m pip install ipykernel

In [21]:
%pip install pymysql

Looking in indexes: https://www.incog.jfrog.ford.com/artifactory/api/pypi/pyserv-native-python-gold-remote/simple/
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [22]:
%pip install pymysql sqlalchemy pandas firebase-admin


Looking in indexes: https://www.incog.jfrog.ford.com/artifactory/api/pypi/pyserv-native-python-gold-remote/simple/
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [23]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [24]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [25]:
#estrae los datos de la tabla movie con una duración mayor a 300 minutos
query=sqla.text("select*" \
"               from movie " \
"               where runtime>300")

#para ver el resultado
pd.read_sql(query,conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,NaN,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [26]:
#mascara booleana Para saber qué tipo de valores hay en la columna de available_globally
ag=sqla.text("SELECT DISTINCT available_globally FROM movie")
pd.read_sql(ag,conn)

#cuenta cuantos datos pertenencen a cada valor de available_globally
c=sqla.text('SELECT available_globally, COUNT(*) FROM movie GROUP BY available_globally')
pd.read_sql(c,conn)

#AVG fue recomendación de la IA para borrar datos nulos y sacar el promedio al mismo tiempo 
c_sin_nulos=sqla.text('SELECT AVG(available_globally=0)*100 FROM movie')
pd.read_sql(c_sin_nulos,conn)

,AVG(available_globally=0)*100
0,83.5717


83.57%

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [27]:
#repitiendo la función que usé arriba para que me dé losdiferentes vslores de locale y número de registros en c/u
idioma=sqla.text('SELECT locale, COUNT(*) FROM movie GROUP BY locale')
pd.read_sql(idioma,conn)



,locale,COUNT(*)
0,NaN,11255
1,en,532


De acuerdo con la tabla del resultado.Solo hay 2 tipos de registos. En (English) con 532 y NaN con 11,255

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [28]:

#title=original_title es máscara boleana para ver si o no son iguales los valores de titulos y ASpara guardar en una columna nueva
R=sqla.text('SELECT locale IS NULL AS locale_nulo, title IS NULL AS title_nulo, original_title IS NULL AS OT_nulo, title=original_title AS titulos_iguales, COUNT(*) AS registros FROM movie GROUP BY locale_nulo, title_nulo, OT_nulo, titulos_iguales')
pd.read_sql(R,conn)

,locale_nulo,title_nulo,OT_nulo,titulos_iguales,registros
0,1,0,1,NaN,7291
1,1,0,0,0.0,3947
2,0,0,1,NaN,532
3,1,0,0,1.0,17


Hay 3947 películas con un título diferente a su título original. La cantidad de títulos diferentes no coincide con la cantidad de valores nulos en locale porque hay 7823 registros faltantes en la columna de títulos originales. No creo que se pueda decir que a ciencia cierta que los valores nulos de locale sean solo de películas en otro idioma por la cantidad de datos faltantes en la colimna de los títulos originales

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [29]:
#MAX()saca los máximos de cumulative weeks in top10. GROUP BY se ocupa para que max funcione. ORDER BY y DESC LIMIT10 ordena y los máximos 10 en orden descendente
M_top10=sqla.text('SELECT m.title, v.movie_id, MAX(v.cumulative_weeks_in_top10) AS Max_tiempo_TOP10 FROM view_summary v JOIN movie m ON v.movie_id=m.id GROUP BY v.movie_id, m.title ORDER BY Max_tiempo_TOP10 DESC LIMIT 10')
pd.read_sql(M_top10,conn)

,title,movie_id,Max_tiempo_TOP10
0,The Boss Baby,11197,22
1,The Super Mario Bros. Movie,11237,21
2,Under Paris,11374,21
3,Despicable Me 2,11161,19
4,Through My Window,177,18
5,Sing,11221,17
6,Paw Patrol: The Movie,11123,15
7,Nowhere,119,15
8,Exterritorial,11154,15
9,Minions,8,14


Un Jefe en Pañales ha permanecido más tiempo en el TOP.

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [30]:
SN=sqla.text('SELECT t.title, s.tv_show_id, COUNT(s.id) AS max_temporadas FROM season s JOIN tv_show t ON s.tv_show_id=t.id GROUP BY s.tv_show_id, t.title ORDER BY max_temporadas DESC LIMIT 5')
pd.read_sql(SN,conn)

,title,tv_show_id,max_temporadas
0,Grey's Anatomy,347,20
1,Naruto Shippuden,1020,20
2,Heartland (2007),564,17
3,It's Always Sunny in Philadelphia,2207,16
4,NCIS,431,15


Los 5 programas de TV con mayor cantidad de temporadas con Grey´s Anatomy, Naruto, Heartland, It´s always sunny in Philadelpia y NCIS.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [31]:
d=sqla.text('SELECT DISTINCT duration FROM view_summary')
pd.read_sql(d,conn)

d2=sqla.text("SELECT DISTINCT start_date, end_date FROM view_summary WHERE duration='SEMI_ANNUALLY' ORDER BY start_date")
pd.read_sql(d2,conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [32]:
GA=sqla.text("SELECT t.title AS serie, s.season_number, s.title AS temporada, v.views FROM view_summary v JOIN season s ON v.season_id = s.id JOIN tv_show t ON s.tv_show_id = t.id WHERE t.title LIKE 'Grey%Anatomy%' AND v.duration = 'SEMI_ANNUALLY' AND v.start_date = '2024-01-01' ORDER BY v.views DESC")
pd.read_sql(GA,conn)

,serie,season_number,temporada,views
0,Grey's Anatomy,1,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy,2,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy,3,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy,5,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy,4,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy,6,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy,7,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy,8,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy,9,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy,10,Grey's Anatomy: Season 10,2400000


Creo que esto logra demostrar que es complejo mantener el interés de la audiencia en el show. Podría ser que la primera temporada la ven muchas más personas porque hay tanto gente que le gustan las series médicas como DR. House o Grey´s anatomy, así como personas que intentan darle una oportunidad a la serie y al final no los termina enganchando, por lo que en la primera temporada se tienen mas views que en las demás. 
Por las views en la temporada 2, yo interpreto que siguen siendo de gente a la que le gustan las series médicas y todavía algo de la gente que se logró enganchar con la primera temporada o quiso darle una oportunidad a la segunda a ver si se pondía mejor. 
Para la tercera vuelve a bajar y mantiene las vistas en algunas temporadas porque ya es gente a la que si le gustó la segie, pero se muestra un patron en el que conforme más avanzan de temporada, abandonan el show. Me atrevería a decir que e spor falta de interés porque no creo que se puedan desarrollar demasiadas tramas en un ambiente hospitalario sin que la serie se vuelva cada vez más tediosa y repetitiva.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [33]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [34]:
#filta el df de tv_show buscando que el texto de la columna diga algo con Grey's anatomy
g=tv_show[tv_show['title'].str.contains("Grey's Anatomy", case=False)]

#junta de df de season y g con el id del tv show
s_df=pd.merge(g,season, left_on='id', right_on='tv_show_id', suffixes=('_show', '_season'))

#junta df de s y de view_summary con el id de season
s_df=pd.merge(s_df, view_summary, left_on='id_season', right_on='season_id')

#filtra el df de s por duración y fecha de inicio. Junta las 2 condiciones con &
s_df=s_df[(s_df['duration']=='SEMI_ANNUALLY') & (pd.to_datetime(s_df['start_date'])=='2024-01-01')]

s_df=(s_df[['title_season', 'season_number', 'views']].sort_values(by='views', ascending=False))

s_df

,title_season,season_number,views
0,Grey's Anatomy: Season 1,1.0,3600000
2,Grey's Anatomy: Season 2,2.0,3100000
4,Grey's Anatomy: Season 3,3.0,2900000
6,Grey's Anatomy: Season 5,5.0,2900000
8,Grey's Anatomy: Season 4,4.0,2900000
10,Grey's Anatomy: Season 6,6.0,2800000
12,Grey's Anatomy: Season 7,7.0,2700000
14,Grey's Anatomy: Season 8,8.0,2600000
16,Grey's Anatomy: Season 9,9.0,2500000
18,Grey's Anatomy: Season 10,10.0,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [35]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [36]:
#from google.colab import drive
#drive.mount('/content/drive')

In [37]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = r"C:\Users\DPEREZ92\Documents\Maestría\1er Trimestre\Ciancia y Analítica de Datos\Tarea 3"
os.chdir(DIR)

In [38]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
fdb = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [43]:
import pandas as pd

# Recorre todos los documentos de la colección EMPLOYEE
docs = fdb.collection('EMPLOYEE').stream()

# doc.id no viene dentro de to_dict(), por eso se agrega a mano
employee = pd.DataFrame([{'id': doc.id, **doc.to_dict()} for doc in docs])

# Deja el id como índice
employee = employee.set_index('id')

employee = employee[['emp_fname', 'emp_lname', 'emp_hiredate']]

employee

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [42]:
firebase_admin.delete_app(firebase_admin.get_app())

ValueError: The default Firebase app does not exist. Make sure to initialize the SDK by calling initialize_app().

No termino de entender por qué sale ese error, pero de acuerdo con IA, este error no afecta el resultado de lo que se pide en las preguntas de arriba.

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---

---

**Declaración de uso de IA**


*   Anthropic. (2026). Claude 5.5 Sonnet [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://claude.ai/ Se utilizó como apoyo para depurar errores de autenticación con Firebase y estructurar el código de la pregunta 10 , así como para encontrar funciones que hicieran más facil lo que se pedía en ciertos pasos. El análisis y la interpretación de resultados son propios.

---